--- a/src/libstore/unix/build/derivation-builder.cc
+++ b/src/libstore/unix/build/derivation-builder.cc
@@ -609,18 +609,25 @@
 }
 
 /* Move/rename path 'src' to 'dst'. Temporarily make 'src' writable if
-   it's a directory and we're not root (to be able to update the
-   directory's parent link ".."). */
+   it's a directory we own or we're not root (to be able to update the
+   directory's parent link ".." even without DAC override capabilities). */
 static void movePath(const std::filesystem::path & src, const std::filesystem::path & dst)
 {
     auto st = lstat(src);
 
-    bool changePerm = (geteuid() && S_ISDIR(st.st_mode) && !(st.st_mode & S_IWUSR));
+    bool changePerm = S_ISDIR(st.st_mode) && !(st.st_mode & S_IWUSR)
+                      && (geteuid() != 0 || st.st_uid == geteuid());
 
     if (changePerm)
         chmod(src, st.st_mode | S_IWUSR);
 
-    std::filesystem::rename(src, dst);
+    try {
+        std::filesystem::rename(src, dst);
+    } catch (...) {
+        if (changePerm)
+            chmod(src, st.st_mode);
+        throw;
+    }
 
     if (changePerm)
         chmod(dst, st.st_mode);
